src/Lecture4.hs

```diff
@@ -223,7 +306,9 @@ implement the next task.
 -}
 
 combineRows :: NonEmpty Row -> Stats
-combineRows = error "TODO"
+combineRows (r :| [])      = rowToStats r
+combineRows (r :| [r'])    = rowToStats r <> rowToStats r'
+combineRows (r :| (r':rs)) = rowToStats r <> combineRows (r' :| rs)
```

Comment:
Without providing detailed feedback yet and taking only a quick look, this function looks like the main problem. This function is not tail-recursive and hence leaks memory.

To understand the issue here better, consider the following implementation of `sum`:

```haskell
sum :: [Int] -> Int
sum [ ] = 0
sum (x : xs) = x + sum xs
```

If you try to evaluate sum `[1 .. 1000]` you will get the following result, according to the implementation:

```haskell
sum [1 .. 1000]
= 1 + sum [2 .. 1000]
= 1 + (2 + sum [3 .. 1000])
= 1 + (2 + (3 + sum [4 .. 1000]))
...
```

and so on. This `sum` function doesn't add numbers immediately. It stores evaluation results in memory until the end of the list. The same is happening with the `combineRows` function. 

This has nothing to do with lazy evaluation, it's just a general problem with not tail-recursive functions in every language 🙂 

To fix this problem, you need to either implement a recursive function like `go` with the bang-patterned accumulator `!acc` or you can implement a separate function that combines non-empty list of `Stats`:

```haskell
combineRows :: NonEmpty Stats -> Stats
```

This function can be easily implemented in one line using fold. And if you use `foldl'` specifically, it will have the desired performance 🙂 

And then, using this function, you can implement `combineRows` in a single line as well 🏎️